c++/src/kj/mutex.c++

```diff
@@ -342,6 +375,10 @@ void Mutex::unlock(Exclusivity exclusivity, Waiter* waiterToSkip) {
       KJ_DASSERT(futex & SHARED_COUNT_MASK, "Unshared a mutex that wasn't shared.");
       uint state = __atomic_sub_fetch(&futex, 1, __ATOMIC_RELEASE);
 
+      if ((state & SHARED_COUNT_MASK) == 0) {
+         releasingShared();
```

Comment:
Technically the reader lock is _already_ released at this point, if the count reached zero, and some other thread could could already be writing a new location (maybe even an exclusive one), making this a bit of a race. Since this is just for debugging I suppose it doesn't matter, but maybe worth a comment. Or maybe you shouldn't bother trying to erase the location info when releasing a shared lock...

Reply:
Hmm... that's a good point. I like your suggestion of not bothering to erase the location info.